### ISSUE TITLE
DOC: Make ruff less rough for new contributors

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#16035]
We took the risk of increasing contribution barrier to have more automatic checks (started with `flake8` replacement but looks like we keep adding stuff to https://github.com/astropy/astropy/blob/main/.ruff.toml ). Now the risk is catching up with us. As mentioned in https://astropy-report.orgmycology.com/sections/community_recommendations#beginner-resources (and also through the document), new contributors are intimidated by `ruff` checks and feels like perfection is killing practical contributions.

*A newsletter or how-to on the new coding checks like ruff would be helpful.*

It is mentioned briefly in https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions but that's about it.

**Comment by MridulS** [comment]
Could reiterating the use of `pre-commit` help here? With `pre-commit` properly installed a new contributor shouldn't need to worry about these automatic checks, and the pre-commit runs ruff with fixes enabled. The ruff formatter in https://github.com/astropy/astropy/pull/16029 would also automatically fix the problems.

Maybe putting `pre-commit` prominently followed by a `git commit` is the answer here?

**Comment by taldcroft** [comment]
This survey result should also serve as a cautionary tale about enforcing type checking, as opposed to encouraging it as a benefit for humans reading the code.

**Comment by MridulS** [comment]
> This survey result should also serve as a cautionary tale about enforcing type checking, as opposed to encouraging it as a benefit for humans reading the code.

I have stayed away from the typing conversations in astropy as I am still pretty new to the community (and I have some strong-ish opinions about adding types to old scientific python code 😄 )

IMO adding types and then **not** enforcing them seems like an anti-pattern. Adding types makes the code very verbose and difficult for beginners to follow along (IMO). If we are paying the cost of adding types, we should at least make sure that these types are being tested and don't drift as people change code.

Adding types just for the sake of IDEs doesn't seem too great to me (again just opinions :) ).
If we do add types it makes sense to exploit all they offer, including strict type checking wherever there are types defined.

**Comment by taldcroft** [comment]
@MridulS - I made a post in the astropy Slack #type-hinting channel. Feel free to express your opinions there, always valued!

**Comment by eerovaher** [comment]
Would it be possible to set up a bot that would post `pre-commit` installation instructions to a pull request if the `pre-commit.ci` check fails?

**Comment by pllim** [comment]
Re: doc and bot -- ¿Por que no los dos?

If those passionate about these solutions have the capacity, does not hurt to have some draft PRs out to explore the different ideas. Thanks! 😄

**Comment by mhvk** [comment]
While I think the extra burdens we add in terms of style, etc., certainly don't help newcomers, I think it may also be the case that there simply is not so much to do for newcomers any more: looking through the 100 currently open PRs, nearly all are bugfixes, style fixes, or refactoring. This is not necessarily bad - it just means that astropy is getting fairly feature complete, like was the case for numpy for a long time already, and that the things that remain (like `uncertainty` and `nddata`) are quite hard. All this makes it a less interesting package for new contributors.

But a bit more on topic, one suggestion would be to adjust the documentation a bit to really encourage people to start by raising an issue (something I found very hard to get even my own students to do for my own package!). In discussing an issue, one often finds out soon enough whether someone is interested in trying to contribute a fix/addition, and then there is already a bit of a relationship with someone who can review/help.

**Comment by dhomeier** [comment]
> While I think the extra burdens we add in terms of style, etc., certainly don't help newcomers, I think it may also be the case that there simply is not so much to do for newcomers any more: looking through the 100 currently open PRs, nearly all are bugfixes, style fixes, or refactoring. This is not necessarily bad - it just means that astropy is getting fairly feature complete, like was the case for numpy for a long time already, and that the things that remain (like `uncertainty` and `nddata`) are quite hard. All this makes it a less interesting package for new contributors.

Looking at _issues_ labelled **feature request** however there are still nearly 400 open, but again very few of them are also labelled **low effort** or **good first issue**. That might reflect a bit of impostor syndrome in labelling, or the fact that we are short in maintainer time to guide and supervise newcomers on such projects. But then I have found that many of the refactoring PRs end up demanding a lot of maintainer time and input as well, because often it turns out to be not as simple as "let pre-commit implement all the Ruff rules".

**Comment by eerovaher** [comment]
There's quite a lot information about using Ruff in #14818, but it would seem that issue is well hidden among the other 1200 open issues we have.

**Comment by nstarman** [comment]
Pin?

**Comment by pllim** [comment]
Why not document what is useful in #14818 in dev docs?

**Comment by dhomeier** [comment]
Yes, this should be in the dev docs – the issue is or should be linked from all relevant PRs, so is not that hidden, but just having the recipes in the description they still seem to be easily overlooked.

**Comment by eerovaher** [comment]
The instructions in #14818 describe addressing currently ignored Ruff rules and are less useful for normal development work. Furthermore, they were written for inexperienced contributors. My idea was that when the simpler Ruff rules are addressed then such detailed instructions will have outlived their usefulness, so there's no need to add them to the developer documentation.

The reason I mentioned #14818 here is that apparently there are new contributors who would like detailed instructions about using Ruff, but they seem to be completely unaware of #14818 despite it explicitly mentioning Ruff in its title and it having the `good first issue` and `Package-novice` labels.

**Comment by pllim** [comment]
OK I pinned #14818 for now... Let's see if that helps in the meantime...
